doc/specifications/proposed/replicator/architecture.rst

```diff
@@ -0,0 +1,142 @@
+==========
+Replicator
+==========
+
+This document describes a common synchronization mechanism between two systems,
+which can cope with write failures, network downtimes and other types of
+partitions. The approach only works with a single write master (Single Source
+of Truth) and any number of read slaves.
+
+In our case we want to use this approach to synchronize the content database
+with a search index.
+
+General Workings
+================
+
+The general scheme for synchronizations is fairly simple:
+
+.. image:: synchronize.png
+
+It generally can be described as a three step process:
+
+1) Fetch last revision from target
+
+   The target is asked for the last revision the target knows. This is the
+   revision of the last successful write operation. For the first request the
+   target may return something like ``null``.
+
+2) Fetch all changes since that last revision from source
+
+   The source (write-master) is asked for all updates since the last revision.
+   A limit, to limit the transfer size, may also be applied. The results
+   **must** be sorted by their revision.
+
+3) Commit all updates on target
+
+   All fetched updates are committed on the target. The target **must not**
+   store a revision if the corresponding update did not succeed.
+
+   This is usually possible to implement using transactions. Another way is to
+   store the revision directly in the received update (document). This way, if
+   the document update is lost, the revision will also be "lost".
```

Comment:
A problem:

- with legacy content engine, not all content-editing actions are atomic. E.g. change section of a node or hiding it do affect its children. So if you sync events 1) move child Y of node X away, 2) hide node X, but the sync of event 1 fails for network reasons, and you succeed to sync event 2, the final status is a discrepancy: node Y will be hidden vs. unhidden

- some actions are even harder to replicate over from master to slave in case any single update is lost (e.g. create node X then its child Y, but when syncing the creation of X is lost)